### ISSUE TITLE
Upgrading `http.client` to use default email policy for parsing

### ISSUE THREAD

**Issue by michaelfm1211** [python/cpython#105622]
To parse HTTP headers, the `http.client` module uses the `email.parser.Parser` class, as headers are in the same format. The relevant function for this is [`_parse_header_lines`](https://github.com/python/cpython/blob/main/Lib/http/client.py#L224-L236). Because `http.server` uses `http.client` to parse its headers, this function is also responsible for parsing headers on the server side.

However, `_parse_header_lines` doesn't pass an [email policy](https://docs.python.org/3/library/email.policy.html) to `email.parser.Parser`. The documentation mentions that [a policy should always be provided, and that the default policy will change in the future](https://docs.python.org/3/library/email.parser.html#email.parser.BytesFeedParser), so it's probably a good idea to change the code to use a specific policy.

The current default policy for `email.parser.Parser` is [`email.policy.compat32`](https://docs.python.org/3/library/email.policy.html#email.policy.compat32), which is a bit outdated but provides complete backwards compatibility. Because the documentation says that this won't be the case forever, I suggest that we change `_parse_header_lines` to use [`email.policy.default`](https://docs.python.org/3/library/email.policy.html#email.policy.default) instead. While this is a breaking change, it would greatly extend support for the many extra RFCs that `email.policy.default` already supports.

One of these RFCs in particular that this change would start supporting is RFC 2047, which includes receiving encoded-word header values and would resolve part of issue #105530.

Note: [this is also on discuss.python.org](https://discuss.python.org/t/upgrading-http-client-to-use-default-email-policy-for-parsing/27632).

### Linked PRs
* gh-105918

**Comment by vstinner** [comment]
I close the issue since the PR has been closed.

**Comment by bitdancer** [comment]
I'm not sure why you closed this issue?  This is something that needs to be done, sooner or later.  Although the policy that should be used is the http policy, that's what it's designed for.

**Comment by vstinner** [comment]
I closed the PR and the issue since the issue reporter, who is also the PR author, was idle for 1 year, didn't sign the CLA, and didn't address my review.

**Comment by bitdancer** [comment]
Fair enough.  However, this is something that should be done, so I'd like to keep the issue open.  It will be a while before I could do anything about it, unfortunately.

**Comment by serhiy-storchaka** [comment]
I measured what using `email.policy.HTTP` for parsing does today.

It is `default.clone(linesep='\r\n', max_line_length=None)`, so it inherits `EmailPolicy`'s `header_factory` and applies email's header registry — keyed on RFC 5322 field names, several of which HTTP reuses with different syntax:

| sent | main | `email.policy.HTTP` |
| --- | --- | --- |
| `Date: Mon, 01 Jan 2024 00:00:00 GMT` | unchanged | `Mon, 01 Jan 2024 00:00:00 +0000` |
| `Content-Type: text/html; charset=utf-8` | unchanged | `text/html; charset="utf-8"` |
| `X-Enc: =?utf-8?q?caf=C3=A9?=` | literal | `café` |
| `To: not an address` | `'not an address'` | `'"not an address"'`, 4 × `InvalidHeaderDefect` |

`Date` is no longer a valid HTTP date (RFC 9110 section 5.6.7 requires IMF-fixdate with a literal `GMT`), RFC 2047 encoded-words are not HTTP syntax, and `To` and `Sender` are not address fields in HTTP.

There is no parse-side validation either, so non-token field names (`Foo(bar): 1`), empty field names, NUL in values, whitespace before the colon and duplicate `Content-Length` are all still accepted. Only obs-fold improves: `'5\r\n  , 6'` becomes `'5  , 6'`, as RFC 9112 section 5.2 asks.

So the http policy needs its own header registry rather than email's, plus validation of HTTP field syntax on parse. Swapping it in as-is passes `test_httplib`, `test_httpservers`, `test_urllib`, `test_urllib2`, `test_xmlrpc` and `test_wsgiref`, so none of this is covered today.

**Comment by michaelfm1211** [comment]
> I closed the PR and the issue since the issue reporter, who is also the PR author, was idle for 1 year, didn't sign the CLA, and didn't address my review.

Oops, sorry! Between the time when the posted this PR/issue and your review, my life has moved on to other priorities and I'm no longer involved in software development (for the better, don't worry!). I'm only seeing this now after randomly checking GitHub. I wish I could help out with this issue or finish the PR but I'm no longer well versed in these things and don't understand what I wrote 3 years ago. Best of luck and sorry for any troubles.
